src/main/kotlin/step3/domain/RacingGame.kt

```diff
@@ -0,0 +1,15 @@
+package step3.domain
+
+import step3.model.Car
+
+class RacingGame(
+    private val randomGenerator: RandomGenerator,
+    cars: List<Car>
+) {
+    private val _cars: List<Car> = cars.toList().map { it.copy() }
+    val cars: List<Car> get() = _cars.map { it.copy() }
```

Comment:
프로퍼티 선언부의 _cars는 어떤 목적으로 선언 하신걸까요? Mutable 자료형을 외부에 노출시키지 않는 목적으로 [backing property](https://kotlinlang.org/docs/properties.html#backing-properties)를 만드는 경우가 있긴 하지만 현재의 구조에서는 불필요해보입니다 🙂 만약 방어적 복사를 의도하신거라면 다음 `toList()` 키워드를 제외하고 나머지는 제거해보면 어떨까요?

```suggestion
    cars: List<Car>
)
    val cars: List<Car> = cars.toList()
```

Reply:
1. 외부에서 (외부에서 생성한)cars를 넘겨 RacingGame 객체를 생성한다. (외부에서 생성한)car.goFoward()를 했을 경우 RacingGame의 car에 영향을 줄 것이다.
2. RacingGame의 cars의 car를 외부에서 변경시킬 수 있다. 
~~~
fun main(){
    val cars = listOf(Car(), Car(), Car()) 
    val racingGame = RacingGame(cars)
    cars[0].goForward()  // 1번 사항
    racingGame.cars[0].goForward()  // 2번사항 
}
~~~

이 두가지 측면을 고려했을때 
~~~
val cars: List<Car> = cars.toList()
~~~
이 코드로는 저 사항들을 해결 할 수 없지 않을까? 하는 생각인데 수현님 의견을 들어보고 싶습니다! 🙏 😄

Reply:
변경사항의 전파에 대한 고민 이해했습니다 👍 
1번과 2번 모두 발생하면 안되는 사이드 이펙트라고 생각해요. 그리고 외부에서 RacingGame의 cars를 조회할 때 getter로 조회하도록 구현하신 이유는 2번에 대한 고민의 결과라고 이해했습니다.

저는 여기서 불변과 가변에 대한 관점으로 고민이 이어지는 것 같아요.
getter로 조회하는 cars도 결국엔 읽기 전용이라고 하기에는 어렵습니다. 외부에서 RacingGame 인스턴스의 cars값을 꺼내온 순간 내부의 car를 조작하는 곳이 있다면 결국 이 값은 원본의 데이터를 보장한다고 볼 수 없는 것이죠. 사실 생각해보면 Car 내부의 goForward()가 내부의 데이터를 조작 가능한 가변으로 설계되었다면 Car를 사용하는 다른 곳에 그 가변이 전파됩니다.

```
Car(변경 가능함) <- RacingGame(변경 가능함 - 전파됨) <- main(원본 데이터를 보장할 수 없음)
```

 getter를 사용하는 것은 일시적으로 외부의 변경사항을 내부로 전파하지 않는 좋은 방법이 될 수는 있겠지만 근본적인 문제를 해결하지는 못합니다.

그럼 이런 상황에서 어떤 방법이 해결책이 될 수 있을까요? 현재 가변 객체인 Car를 변경 불가능한 불변 객체로 설계해볼 수 있습니다. 내부의 변경사항이 외부로 전파되지 않도록 설계하는 방법입니다. 

```
Car(변경 불가능) <- RacingGame(변경 불가능) <- main(원본 데이터를 보장할 수 있음)
```

Car 불변 객체로 만들면 게임이 아예 불가능하지 않나? 어떻게 값을 변경하라는 거지...? 라는 의문이 들 수 있습니다. Kotlin Collection의 add/remove 함수와 plus/minus 함수의 차이를 보시면 불변 객체에서 값을 변경하는 것을 어떻게 트래킹하는지 이해하실 수 있을거예요!

이야기가 길어졌네요! 길게 말씀드렸지만 불변함에 대한 고민은 자동차 경주 미션에서 의도하는 목표와는 조금 다르기에 지금 단계에서는 고민해보고 넘어가는 것으로도 충분할 것 같습니다 :) 로또 미션 이후부터 이러한 고민을 또 하시게 될 텐데 나중에 참고해보세요 🙂